week1/tictactoe/borboss366/Main.hs

```diff
@@ -0,0 +1,353 @@
+
+calcWinningState :: TTTPlay -> (TTTPlay -> [TTTRow]) -> TTTWinner
+calcWinningState tttP rowFunc | isJust wr = (snd . fromJust) wr
+                              | calcTotalFields cnf == movesCount tttP = Deuce
+                              | otherwise = UnknownWinner
+                              where
+                                cnf = config tttP
+                                wr = calcWinningRow tttP rowFunc
+
+calcWinningRow :: TTTPlay -> (TTTPlay -> [TTTRow]) -> Maybe (TTTRow, TTTWinner)
+calcWinningRow tttP rowFunc
+  | null winningRows = Nothing
+  | otherwise = Just (head winningRows)
+  where
+    cnf = config tttP
+    winningRows = filter (isW.snd) $ map (\row -> (row, checkRowWinner cnf row)) (rowFunc tttP)
+    isW (TTTWinner _) = True
+    isW _ = False
+
+
+calcAllWinningState tttp = calcWinningState tttp allRows
+
+calcMoves :: TTTPlay -> TTTStepType -> Int
+calcMoves tttP tttSt = length $ filter (\t -> stepType t == tttSt) (steps tttP)
+
+isFinishedWinner :: TTTWinner -> Bool
+isFinishedWinner (TTTWinner Krestig) = True
+isFinishedWinner (TTTWinner Nolek) = True
+isFinishedWinner Deuce = True
+isFinishedWinner _ = False
+
+isFinished :: TTTPlay -> Bool
+isFinished tttP = isFinishedWinner $ calcAllWinningState tttP
+
+calcNextAct :: TTTPlay -> Maybe TTTStepType
+calcNextAct tttP  | isFinished tttP = Nothing
+                  | stK > stN = Just Nolek
+                  | stK < stN = Just Krestig
+                  | otherwise = Just $ firstStep $ config tttP
+                  where
+                    stK = calcMoves tttP Krestig
+                    stN = calcMoves tttP Nolek
+
+{-- Parsing functions --}
+parseX :: String -> Maybe Int
+parseX str  | length str == 1 && isAsciiLower c = Just $ ord c - ord 'a'
+            | otherwise = Nothing
+            where c = toLower $ head str
+
+parsePosition :: String -> Maybe TTTPosition
+parsePosition line  | length vals == 2 = let
+                        xm = parseX (vals !! 0)
+                        ym = (+ (-1)) <$> readMaybe (vals !! 1) :: Maybe Int
+                      in pure TTTPosition <*> xm <*> ym
+                    | otherwise = Nothing
+                    where g c1 c2 = isDigit c1 == isDigit c2
+                          vals = groupBy g line
+
+parseConfig :: String -> Maybe TTTConfig
+parseConfig line  | length values == 3 && all isJust values =
+                    let [w,h,row] = map fromJust values in
+                      Just TTTConfig {
+                        fieldWidth = w,
+                        fieldHeight = h,
+                        inRow2Win = row,
+                        firstStep = Krestig
+                      }
+                  | otherwise = Nothing
+                  where values = map (readMaybe :: String -> Maybe Int) (words line)
+
+readUntilGood :: String -> (String -> Maybe a) -> IO a
+readUntilGood title parseFun = do
+  putStrLn title
+  line <- getLine
+  let object = parseFun line
+  if (isJust object)
```

Comment:
Instead of using `isJust` everywhere, it's better to pattern-match:

``` haskell
case parseFun line of
  Just obj -> return obj
  Nothing  -> readUntilGood title parseFun
```